### PR TITLE
Implement the line-height calculation quirk

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11853)

### CODE REVIEW

**Comment by jdm** [comment]
@notriddle Would you be interested in reviewing this change?

**Comment by notriddle** [comment]
@jdm Sure.

File: components/layout/inline.rs [hunk 1]
```diff
@@ -1798,6 +1823,16 @@ fn inline_contexts_are_equal(inline_context_a: &Option<InlineFragmentContext>,
...
 }
 
+fn side_borders_and_paddings_are_definitely_zero(style: &ServoComputedValues) -> bool {
+    let border = style.get_border();
+    let padding = style.get_padding();
+
+    border.border_left_width == Au(0) &&
+    border.border_right_width == Au(0) &&
+    padding.padding_left.is_definitely_zero() &&
+    padding.padding_right.is_definitely_zero()
```
Comment (notriddle) [inline components/layout/inline.rs]: `is_definitely_zero` is, as the name implies, not an exhaustive test for whether the "used property" of a value is zero. The spec asks for the used value.

How hard would it be to get the used value here?

**Comment by notriddle** [comment]
Other than that and not having any tests, LGMT.

**Comment by notriddle** [comment]
Looking through the IRC backlog, it looks like the used value only exists after layout. So no code changes needed, just adding a reftest should work.

**Comment by Ms2ger** [comment]
The _definitely_ bits scare me. Can we produce a test where the used value is zero, but not definitely zero?

**Comment by nox** [comment]
It isn't scary, it just means it won't catch cases where the padding is `calc(10% - 10px)` and the containing block is 100px wide. I'm not even sure other UAs support that, it goes against the spirit of the spec (IMO), and Servo is definitely unable to get the used value at that point in layout, we would need to call `compute_minimum_ascent_and_descent` in a later layout pass and for now @mbrubeck told me to just ignore that case.

**Comment by Ms2ger** [comment]
If we willfully ignore the spec, we should file an issue.

**Comment by notriddle** [comment]
If none of the browsers implement the spec as written, we should file a bug against the *spec*.

**Comment by notriddle** [comment]
*[components/layout/inline.rs, line 1833 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11853#-KL1YinYqjNFsSpkVdtg-r1-1833:-KL24kt97xDKCSJufjOB:2007948272) ([raw file](https://github.com/servo/servo/blob/f1d1c1af9ea02f86270bc49ac9f922bbc6a7df46/components/layout/inline.rs#L1833)):*
<details><summary><i>Previously, notriddle (Michael Howell) wrote…</i></summary>
> `is_definitely_zero` is, as the name implies, not an exhaustive test for whether the "used property" of a value is zero. The spec asks for the used value.
> 
> 
> 
> How hard would it be to get the used value here?
</details>
OK, guess getting a truly used value would be too much of a hassle, and probably not web compatible anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11853)*

**Comment by notriddle** [comment]
Reviewed 1 of 3 files at r1.
Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11853)*

**Comment by notriddle** [comment]
Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11853)*

**Comment by emilio** [comment]
I don't think we're deviating from the spec. The definition of used value is:

> The used value is the result of taking the computed value and completing any remaining 
> calculations to make it the absolute theoretical value used in the layout of the document. If the 
> property does not apply to this element, then the element has no used value for that property.

So I don't think we need the specified value at all here, the computed value is more than enough.

**Comment by nox** [comment]
@emilio The problem is that the used value is "more computed" than the computed values, and that `calc()` values maybe shouldn't be subject to the quirk.

**Comment by emilio** [comment]
Ahh, I get it, sorry then... Yeah, this wouldn't catch less-tricky as "computed value is 10%", but width is 0 thing, right?

In that case I agree with @Ms2ger, a spec bug should be open at least, though given how servo does layout I doubt it's possible to fix this easily.

**Comment by zcorpan** [comment]
This quirk should also be present in limited-quirks mode.

File: components/layout/inline.rs [hunk 2]
```diff
@@ -1249,6 +1265,15 @@ impl InlineFlow {
...
         }
     }
+
+    fn contains_text_fragments(&self) -> bool {
+        self.fragments.fragments.iter().any(|fragment| {
+            match fragment.specific {
+                SpecificFragmentInfo::ScannedText(_) => true,
+                _ => false,
+            }
+        })
```
Comment (pcwalton) [inline components/layout/inline.rs]: nit: How about `self.fragments.fragments.iter().any(Fragment::is_scanned_text_fragment)`?

File: components/layout/inline.rs (hunk 1 above)
Comment (pcwalton) [inline components/layout/inline.rs]: Shouldn't be hard; use the `self.border_padding` field.

**Comment by jdm** [comment]
It's not clear to me from reading the previous comments on what the status is here. Are we waiting on some other piece of work, or spec discussion? I seem to recall @nox mentioning that these changes were incomplete in some way, but I might be misremembering. Who's responsible for taking the next step here?

**Comment by notriddle** [comment]
["The fix is completely wrong."](http://logs.glob.uno/?c=mozilla%23servo&s=27+Jun+2016&e=27+Jun+2016#c463425) Either he fixes it or it goes back into the big pot of unassigned issues.

**Comment by Ms2ger** [comment]
`<nox> Ms2ger: Go ahead and close it.`

**Pull request closed without merging**